cli/src/main.rs

```diff
@@ -234,6 +229,57 @@ pub(crate) struct CoreOptions {
     core: usize,
 }
 
+#[derive(clap::Parser)]
+pub(crate) struct FormatOptions {
+    #[clap(value_enum, ignore_case = true, default_value = "elf", long)]
+    format: Format,
+    /// The address in memory where the binary will be put at.
+    #[clap(long)]
+    pub base_address: Option<u64>,
+    /// The number of bytes to skip at the start of the binary file.
+    #[clap(long, default_value = "0")]
+    pub skip: u32,
+    /// The idf bootloader path
+    #[clap(long)]
+    pub bootloader: Option<PathBuf>,
+    /// The idf partition table path
+    #[clap(long)]
+    pub partition_table: Option<PathBuf>,
+}
+
+impl FormatOptions {
+    pub fn into_format(self) -> anyhow::Result<Format> {
+        Ok(match self.format {
+            Format::Bin(_) => Format::Bin(BinOptions {
+                base_address: self.base_address,
+                skip: self.skip,
+            }),
+            Format::Hex => Format::Hex,
+            Format::Elf => Format::Elf,
+            Format::Idf(_) => {
+                let bootloader = if let Some(path) = self.bootloader {
+                    Some(std::fs::read(path)?)
```

Comment:
What's the reason for reading this here?

I think it would be more consistent to read this in the `load_idf_data` function, where the main binary itself is read as well.

Reply:
The main reason I did this is so that the `IdfOptions` type inside probe-rs itself would be "typed". If I switch to doing the reading inside probe-rs itself `IdfOptions` no longer conveys what the actual types are and would just be paths instead.

```rust
/// Extended options for flashing a ESP-IDF format file.
#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone, Default)]
pub struct IdfOptions {
    /// The bootloader
    pub bootloader: Option<Vec<u8>>,
    /// The partition table
    pub partition_table: Option<esp_idf_part::PartitionTable>,
}
```

versus

```rust
/// Extended options for flashing a ESP-IDF format file.
#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone, Default)]
pub struct IdfOptions {
    /// The bootloader
    pub bootloader: Option<PathBuf>,
    /// The partition table
    pub partition_table: Option<PathBuf>,
}
```

If you see no benefit to this, I can change it to the latter.

Reply:
@Tiwalun any thoughts on my reasoning?

Reply:
Still feels inconsistent to me, but maybe that just means that parsing the main binary is done at the wrong place :) I'm fine with merging it like this.